Adding metrics for GANs evaluation

szmazurek opened the pull request:
Fixes #812

## Proposed Changes
- currently implemented: FID, SSIM, LIPIS
- LPIPS compatibile with 2D and 3D, FID only for 2D

## Checklist

- [X] [`CONTRIBUTING`](https://github.com/mlcommons/GaNDLF/blob/master/CONTRIBUTING.md) guide has been followed.
- [X] PR is based on the [current GaNDLF master ](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/keeping-your-local-repository-in-sync-with-github/syncing-your-branch-in-github-desktop?platform=windows).
- [X] Non-breaking change (does **not** break existing functionality): provide **as many** details as possible for _any_ breaking change.
- [X] Function/class source code documentation added/updated (ensure `typing` is used to provide type hints, including and not limited to using `Optional` if a variable has a pre-defined value).
- [X] Code has been [blacked](https://github.com/psf/black#usage) for style consistency and linting.
- [X] If applicable, version information [has been updated in GANDLF/version.py](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/version.py).
- [X] If adding a git submodule, add to list of exceptions for black styling in [pyproject.toml](https://github.com/mlcommons/GaNDLF/blob/master/pyproject.toml) file.
- [ ] [Usage documentation](https://github.com/mlcommons/GaNDLF/blob/master/docs) has been updated, if appropriate.
- [ ] Tests added or modified to [cover the changes](https://app.codecov.io/gh/mlcommons/GaNDLF); if coverage is reduced, please give explanation.
- [ ] If customized dependency installation is required (i.e., a separate `pip install` step is needed for PR to be functional), please ensure it is reflected in all the files that control the CI, namely: [python-test.yml](https://github.com/mlcommons/GaNDLF/blob/master/.github/workflows/python-test.yml), and all docker files [[1](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-CPU),[2](https://github.com/mlcommons/GaNDLF/blob/devcontainer_build_fix/Dockerfile-CUDA11.6),[3](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-ROCm)].

szmazurek committed:
Adding metrics for GANs evaluation

- currently implemented: FID, SSIM, LIPIS
- LPIPS compatibile with 2D and 3D, FID only for 2D

sarthakpati commented:
There is duplication from here: https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/metrics/synthesis.py. This is the base we are using for the `generate_metrics` module [[ref](https://mlcommons.github.io/GaNDLF/usage/#generate-metrics)].

I would recommend extending this module with additional metrics, and calling from here as needed. Additionally, once you add a metric, the corresponding unit test should be a simple extension of this [[ref](https://github.com/mlcommons/GaNDLF/blob/master/testing/test_full.py#L3051)].

szmazurek commented:
> There is duplication from here: https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/metrics/synthesis.py. This is the base we are using for the `generate_metrics` module [[ref](https://mlcommons.github.io/GaNDLF/usage/#generate-metrics)].
> 
> I would recommend extending this module with additional metrics, and calling from here as needed. Additionally, once you add a metric, the corresponding unit test should be a simple extension of this [[ref](https://github.com/mlcommons/GaNDLF/blob/master/testing/test_full.py#L3051)].

Sure, will move it tomorrow

szmazurek committed:
Merged GAN metrics with synthesis module

szmazurek committed:
Adding second implementation for SSIM

- current synthesis implementation of ssim was giving errors when used
  with no mask
- did not want to break the current synthesis functionality as it will
  probably disrupt your code
- added my implementation of ssim, compatibile with gan training

sarthakpati commented on GANDLF/metrics/gan_utils/fid.py:
```diff
@@ -0,0 +1,524 @@
+        features_list: List[str],
+        feature_extractor_weights_path: Optional[str] = None,
+    ) -> None:
+        if not _TORCH_FIDELITY_AVAILABLE:
+            raise ModuleNotFoundError(
+                "NoTrainInceptionV3 module requires that `Torch-fidelity` is installed."
+                " Either install as `pip install torchmetrics[image]` or `pip install torch-fidelity`."
+            )
```
Change to `assert`.

szmazurek replied:
done

sarthakpati commented on GANDLF/metrics/gan_utils/fid.py:
```diff
@@ -0,0 +1,524 @@
+        super().__init__(**kwargs)
+        if isinstance(feature, int):
+            num_features = feature
+            if not _TORCH_FIDELITY_AVAILABLE:
+                raise ModuleNotFoundError(
+                    "FrechetInceptionDistance metric requires that `Torch-fidelity` is installed."
+                    " Either install as `pip install torchmetrics[image]` or `pip install torch-fidelity`."
+                )
```
Change to `assert`.

szmazurek replied:
done

sarthakpati commented on GANDLF/metrics/gan_utils/fid.py:
```diff
@@ -0,0 +1,524 @@
+                    " Either install as `pip install torchmetrics[image]` or `pip install torch-fidelity`."
+                )
+            valid_int_input = (64, 192, 768, 2048)
+            if feature not in valid_int_input:
+                raise ValueError(
+                    f"Integer input to argument `feature` must be one of {valid_int_input}, but got {feature}."
+                )
```
Change to `assert`.

szmazurek replied:
done

sarthakpati commented on GANDLF/metrics/gan_utils/fid.py:
```diff
@@ -0,0 +1,524 @@
+                name="inception-v3-compat", features_list=[str(feature)]
+            )
+
+        elif isinstance(feature, Module):
+            self.inception = feature
+            if hasattr(self.inception, "num_features"):
+                num_features = self.inception.num_features
+            else:
+                dummy_image = torch.randint(
+                    0, 255, (1, 3, 299, 299), dtype=torch.uint8
+                )
+                num_features = self.inception(dummy_image).shape[-1]
+        else:
+            raise TypeError("Got unknown input to argument `feature`")
```
Change to `assert`.

szmazurek replied:
done

sarthakpati commented on GANDLF/metrics/gan_utils/fid.py:
```diff
@@ -0,0 +1,524 @@
+        else:
+            raise TypeError("Got unknown input to argument `feature`")
+
+        if not isinstance(reset_real_features, bool):
+            raise ValueError(
+                "Argument `reset_real_features` expected to be a bool"
+            )
```
Change to `assert`.

szmazurek replied:
done

sarthakpati commented on GANDLF/metrics/gan_utils/fid.py:
```diff
@@ -0,0 +1,524 @@
+            )
+        self.reset_real_features = reset_real_features
+
+        if not isinstance(normalize, bool):
+            raise ValueError("Argument `normalize` expected to be a bool")
```
Change to `assert`.

szmazurek replied:
done

sarthakpati commented on GANDLF/metrics/gan_utils/fid.py:
```diff
@@ -0,0 +1,524 @@
+
+    def compute(self) -> Tensor:
+        """Calculate FID score based on accumulated extracted features from the two distributions."""
+        if (
+            self.real_features_num_samples < 2
+            or self.fake_features_num_samples < 2
+        ):
+            raise RuntimeError(
+                "More than one sample is required for both the real and fake distributed to compute FID"
+            )
```
Change to `assert`.

szmazurek replied:
done

sarthakpati commented on GANDLF/metrics/gan_utils/functional/lpips.py:
```diff
@@ -0,0 +1,366 @@
+        Tuple[Tensor, Union[int, Tensor]]: loss and total number of images
+
+    """
+    if not (_valid_img(img1, normalize) and _valid_img(img2, normalize)):
+        raise ValueError(
+            "Expected both inputs to be 4D or 5D with shape "
+            "[N x C x H x W] or [N x C x D x H x W]"
+            f"Got input with shape {img1.shape} and {img2.shape}."
+            f" {[img1.min(), img1.max()]} and {[img2.min(), img2.max()]} "
+            f" when all values are expected to be in the "
+            f"{[0,1] if normalize else [-1,1]} range."
+        )
```
Change to `assert`.

szmazurek replied:
done

sarthakpati commented on GANDLF/metrics/gan_utils/functional/lpips.py:
```diff
@@ -0,0 +1,366 @@
+        Union[None, SoftACSConverter, ACSConverter, Conv3dConverter]: the
+    converter to use.
+    """
+    if converter_type is None or converter_type == "soft":
+        converter = SoftACSConverter
+    elif converter_type == "acs":
+        converter = ACSConverter
+    elif converter_type == "conv3d":
+        converter = Conv3dConverter
+    else:
+        raise ValueError(f"Unknown converter type {converter_type}")
```
Change to `assert`.

szmazurek replied:
done

sarthakpati commented on GANDLF/metrics/gan_utils/lpip.py:
```diff
@@ -0,0 +1,150 @@
+        super().__init__(**kwargs)
+
+        valid_net_type = ("vgg", "alex", "squeeze")
+        if net_type not in valid_net_type:
+            raise ValueError(
+                f"Argument `net_type` must be one of {valid_net_type}, but got {net_type}."
+            )
```
Change to `assert`.

szmazurek replied:
done

sarthakpati commented on GANDLF/metrics/gan_utils/lpip.py:
```diff
@@ -0,0 +1,150 @@
+        if reduction not in valid_reduction:
+            raise ValueError(
+                f"Argument `reduction` must be one of {valid_reduction}, but got {reduction}"
+            )
```
Change to `assert`.

sarthakpati commented on GANDLF/metrics/gan_utils/lpip.py:
```diff
@@ -0,0 +1,150 @@
+        if not isinstance(normalize, bool):
+            raise ValueError(
+                f"Argument `normalize` should be an bool but got {normalize}"
+            )
```
Change to `assert`.

sarthakpati commented on GANDLF/metrics/synthesis.py:
```diff
@@ -224,3 +228,376 @@ def ncc_min(prediction: torch.Tensor, target: torch.Tensor) -> float:
+    if params["model"]["dimension"] != 2:
+        raise ValueError("FID is only supported for 2D images")
+    if params["batch_size"] == 1:
+        raise ValueError("FID is not supported for batch size 1")
```
Change to `assert`.

sarthakpati commented on GANDLF/metrics/synthesis.py:
```diff
@@ -224,3 +228,376 @@ def ncc_min(prediction: torch.Tensor, target: torch.Tensor) -> float:
+    if params["model"]["dimension"] != 2:
+        raise ValueError("FID is only supported for 2D images")
+    if params["batch_size"] == 1:
+        raise ValueError("FID is not supported for batch size 1")
```
Change to `assert`.

sarthakpati commented on GANDLF/metrics/gan_utils/functional/lpips.py:
```diff
@@ -0,0 +1,366 @@
+        Tuple[Tensor, Union[int, Tensor]]: loss and total number of images
+
+    """
+    if not (_valid_img(img1, normalize) and _valid_img(img2, normalize)):
```
Put an assert for 2D image check before this line to make it explicit that this module only works for 2D images.

szmazurek replied:
Hmm, but that would not be true - we have lpips working also for 3D images.

sarthakpati commented on GANDLF/metrics/gan_utils/functional/lpips.py:
```diff
@@ -0,0 +1,366 @@
+    Returns:
+        bool: whether input is valid
+    """
+    value_check = value_check = (
+        img.max() <= 1.0 and img.min() >= 0.0 if normalize else img.min() >= -1
+    )
+    return (img.ndim == 4 or img.ndim == 5) and value_check
```
This seems to be same/similar to the check in FID. Can they be combined?

szmazurek replied:
Hmm, if fact maybe, but some considerations - we would have to add a flag that indicates whether we are using it for fid or lpip metrics and add conditionals (as the checks for fid and lpips are different) - wouldn't this make the function more complex and hard to debug? I may write similar for fid (currently fid does similar thing but using separate checks in different places.)

sarthakpati replied:
I think all these checks can be put in a `utils` submodule under this one. Also, would `metrics` be a better name for this module rather than `functional`?

szmazurek replied:
Not really sure what do you mean with "under this one"? You mean GANDLF/utils? With the naming - yeah, we can go this way, but isn't that going to be confusing when in metrics folder we have gan_utils and in this one we would have another metrics folder? Also agree that functional is not really descriptive, I followed torchmetrics style of code organisation.

sarthakpati replied:
Ah okay. This is no longer needed.

sarthakpati commented on GANDLF/metrics/gan_utils/functional/lpips.py:
```diff
@@ -0,0 +1,366 @@
+        Tuple[Tensor, Union[int, Tensor]]: loss and total number of images
+
+    """
+    if not (_valid_img(img1, normalize) and _valid_img(img2, normalize)):
```
Put an assert for 2D image check before this line to make it explicit that this module only works for 2D images.

szmazurek replied:
Same as above

sarthakpati commented on GANDLF/metrics/gan_utils/lpip.py:
```diff
@@ -0,0 +1,150 @@
+from .functional import (
+    lpips_compute,
+    lpips_update,
+    determine_converter,
```
This function is a duplicate of what is there in `GANDLF.models.ModelBase`. Let's keep a note of it so that we combine them when we are merging upstream.

szmazurek replied:
noted

sarthakpati commented on GANDLF/metrics/synthesis.py:
```diff
@@ -224,3 +228,376 @@ def ncc_min(prediction: torch.Tensor, target: torch.Tensor) -> float:
+        torch.Tensor: The FID score.
+    """
+
+    def _get_features_size(params: Dict[str, Any]) -> int:
```
Put an assert for 2D image check before this line to make it explicit that this module only works for 2D images.

szmazurek replied:
added

sarthakpati reviewed (changes requested):
Requesting a few changes:

- Change `raise` to `assert` - code coverage and consistency.
- I may I have requested the 2D image check at multiple places. They should be present in both places, i.e., where the module is being defined (say `GANDLF/metrics/gan_utils/fid`) and as well as where they are being invoked (say `GANDLF/metrics/synthesis`). This will ensure that the `assert` gets triggered regardless of where the user is invoking the code from.

szmazurek committed:
Modifications to fid utils

- changed raises into assertions
- small docstring update

szmazurek committed:
Modifying raise to assert in lpips related code

szmazurek committed:
Fixes for the lpips and synthesis modules

- added new assertions instead of raise statements

sarthakpati commented on GANDLF/metrics/gan_utils/fid.py:
````diff
@@ -0,0 +1,492 @@
+        alternatively we will pass through tensor of shape ``(1, 3, 299, 299)`` and dtype ``torch.uint8``` to the
+        forward pass and expect a tensor of shape ``(1, num_features)`` as output.
+
+    Raises:
+        AssertionError:
+            If torch version is lower than 1.9
+        AssertionError:
+            If ``feature`` is set to an ``int`` (default settings) and ``torch-fidelity`` is not installed
+        AssertionError:
+            If ``feature`` is set to an ``int`` not in [64, 192, 768, 2048]
+        AssertionError:
+            If ``feature`` is not an ``str``, ``int`` or ``torch.nn.Module``
+        AssertionError:
+            If ``reset_real_features`` is not an ``bool``
````
I am guessing this is redundant, now?

szmazurek replied:
Wanted to keep it to tell which errors we are going to catch explicitly, nevertheless I believe you are right - can be removed.

sarthakpati commented on GANDLF/metrics/gan_utils/fid.py:
```diff
@@ -0,0 +1,492 @@
+        Returns:
+            Figure and Axes object
+
+        Raises:
+            ModuleNotFoundError:
+                If `matplotlib` is not installed
```
I am guessing this is redundant, now?

szmazurek replied:
Actually, entire method is. Removing

sarthakpati commented on GANDLF/metrics/gan_utils/lpip.py:
```diff
@@ -0,0 +1,145 @@
+        Returns:
+            Figure and Axes object
+
+        Raises:
+            ModuleNotFoundError:
+                If `matplotlib` is not installed
```
I am guessing this is redundant, now?

szmazurek replied:
As above, entire method redundant and removed

sarthakpati commented on GANDLF/metrics/synthesis.py:
```diff
@@ -224,3 +228,299 @@ def ncc_min(prediction: torch.Tensor, target: torch.Tensor) -> float:
+    information.
+    Returns:
+        torch.Tensor: The SSIM score.
+    """
```
Update comment style

sarthakpati commented on GANDLF/metrics/synthesis.py:
```diff
@@ -224,3 +228,299 @@ def ncc_min(prediction: torch.Tensor, target: torch.Tensor) -> float:
+        n_input_channels (int): The number of input channels.
+    Returns:
+        torch.Tensor: The FID score.
+    """
```
Update comment style

sarthakpati commented on GANDLF/metrics/synthesis.py:
```diff
@@ -224,3 +228,299 @@ def ncc_min(prediction: torch.Tensor, target: torch.Tensor) -> float:
+    real_images: torch.Tensor,
+    params: Dict[str, Any],
+) -> torch.Tensor:
+    """This function computes the LPIPS between the generated images and the
+    real images. Except for the params specified below, the rest of the params
+    are default from torchmetrics.
+    Args:
+        generated_images (torch.Tensor): The generated images.
+        real_images (torch.Tensor): The real images.
+        params (dict): The parameter dictionary containing training and data
+    Returns:
+        torch.Tensor: The LPIP score.
+    """
```
Update comment style.

szmazurek replied:
Updated mentioned and other comment styles - indeed the spacing was wrong there

sarthakpati reviewed (changes requested):
Please ensure that the comment style is consistent for all functions/classes, and I am guessing `Raises` is no longer needed in the comments after everything has been converted to `assert`...

szmazurek committed:
Docstring updates

- updated incorrectly styled docstrings
- added some missing docstrings in inner functions

sarthakpati commented on GANDLF/metrics/gan_utils/fid.py:
```diff
@@ -0,0 +1,428 @@
+    r"""Compute adjusted version of `Fid Score`_.
+
+    The Frechet Inception Distance between two multivariate Gaussians X_x ~ N(mu_1, sigm_1)
+    and X_y ~ N(mu_2, sigm_2) is d^2 = ||mu_1 - mu_2||^2 + Tr(sigm_1 + sigm_2 - 2*sqrt(sigm_1*sigm_2)).
```
```suggestion
    """
    This function calculates the adjusted version of `Fid Score`_. The Frechet Inception Distance between two multivariate Gaussians X_x ~ N(mu_1, sigm_1)
    and X_y ~ N(mu_2, sigm_2) is d^2 = ||mu_1 - mu_2||^2 + Tr(sigm_1 + sigm_2 - 2*sqrt(sigm_1*sigm_2)).
```

sarthakpati commented on GANDLF/metrics/gan_utils/fid.py:
```diff
@@ -0,0 +1,428 @@
+    r"""Calculate Fréchet inception distance (FID_) which is used to access the quality of generated images.
+
+    .. math::
+        FID = \|\mu - \mu_w\|^2 + tr(\Sigma + \Sigma_w - 2(\Sigma \Sigma_w)^{\frac{1}{2}})
```
```suggestion
    """
    This function calculates the Fréchet inception distance (FID_) which is used to access the quality of generated images.
    .. math::
        FID = \|\mu - \mu_w\|^2 + tr(\Sigma + \Sigma_w - 2(\Sigma \Sigma_w)^{\frac{1}{2}})
```

sarthakpati reviewed (changes requested):
The comment style needs to be updated (see some of my example suggestions for reference) for consistency.

szmazurek committed:
Update GANDLF/metrics/gan_utils/fid.py

Co-authored-by: Sarthak Pati <patis@iu.edu>

szmazurek committed:
Update GANDLF/metrics/gan_utils/fid.py

Co-authored-by: Sarthak Pati <patis@iu.edu>

szmazurek committed:
Fixing docstrings

- fixed remaining docstrings in the GAN specific metrics

szmazurek committed:
Black formatting

- formatted the files with new black config

szmazurek commented:
Hey, I belive the full style cleaning was done with the yesterday commit. If you do not find any other issues I believe we can merge.

sarthakpati reviewed (approved):
LGTM
